src/index.js

```diff
@@ -19,664 +20,520 @@ const debug = (function () {
+    masqAppBaseUrl: context.config.masqAppBaseUrl ? context.config.masqAppBaseUrl : jsonConfig.masqAppBaseUrl,
+    swarmConfig: context.config.swarmConfig ? context.config.swarmConfig : jsonConfig.swarmConfig
+  }
 
-              default:
-                // TODO change error
-                throw new MasqError(MasqError.WRONG_MESSAGE, `Unexpectedly received message with type ${json.msg}`)
-            }
-            resolve()
-          } catch (e) {
-            await this._resetLogin()
-            reject(e)
-          }
-        })
+  context = initNullAttributes(context)
 
-        this.loginSwDisconnectListener = async (peer, id) => {
-          debug('[masq.logIntoMasq] disconnect')
-          reject(new MasqError(MasqError.DISCONNECTED_DURING_LOGIN))
-        }
-        this.loginSw.on('disconnect', this.loginSwDisconnectListener)
+  // try to login from session info
+  // await this._loadSessionInfo()
 
-        this.loginSw.on('close', () => {
-          debug('[masq.logIntoMasq] close')
-        })
-      })
-    })
-
-    try {
-      await pr
-      this.loginSw.close()
-    } catch (e) {
-      await this._resetLogin()
-      throw e
-    }
-  }
-
-  _removeDisconnectListener () {
-    if (this.loginSw) {
-      // remove disconnect listener
-      if (this.loginSwDisconnectListener) {
-        this.loginSw.removeListener('disconnect', this.loginSwDisconnectListener)
-        this.loginSwDisconnectListener = null
-      }
-    }
-  }
+  assign(context)
+  await _genLoginLink(context)
+}
 
-  async _resetLogin (keepLoginLink = false) {
-    debug('[masq._resetLogin] keep login link: ' + Boolean(keepLoginLink))
-    this._deleteSessionInfo()
+const initNullAttributes = (context) => {
+  // login state data
+  context.stayConnected = false
+  context.loginSw = null
+  context.loginPeer = null
+
+  // reset replication variables
+  context.replicating = false
+  context.userAppRepSW = null
+  context.userAppRepHub = null
+
+  // reset logged in user variable
+  context.userAppDbId = null
+  context.userAppDb = null
+  context.userAppDEK = null
+  context.importedUserAppDEK = null
+  context.userAppNonce = null
+  context.username = null
+  context.profileImage = null
+
+  return context
+}
 
-    // remove disconnect listener
-    // close login swarm
-    if (this.loginSw) {
-      this._removeDisconnectListener()
+const _startLogin = (context, event, actionMeta) => (callbackParent, onEvent) => {
+  // make stayConnected a boolean
+  context.stayConnected = !!event.stayConnected
+
+  // Subscribe to channel for a limited time to sync with masq
+  debug(`Creation of a hub with ${context.loginChannel} channel name`)
+
+  const hub = signalhub(context.loginChannel, context.config.hubUrls)
+  hub.on('error', ({ url, error }) => {
+    send('SIGNALLING_SERVER_ERROR')
+    // reject(new MasqError(MasqError.SIGNALLING_SERVER_ERROR, url, error))
+  })
+
+  context.loginSw = swarm(hub, {
```

Comment:
I think we should avoid affecting context.loginSw directly. I guess context passed as argument is a copy, but still

Reply:
To be resolved in a future PR